src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs

```diff
@@ -551,19 +551,24 @@ private static string EntryFromPath(string entry, int offset, int length)
             if (length == 0)
                 return String.Empty;
 
-            // create a mutable copy
-            char[] chars = entry.ToCharArray(offset, length);
-
-            // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
-            // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
-            // explicitly trying to standardize to '/'
-            for(int i = 0; i < chars.Length; i++)
+            // create a copy of the string we can modify
+            string result = String.Copy(entry);
+            
+            fixed (char* pChar = result)
             {
-                if (chars[i] == Path.DirectorySeparatorChar || chars[i] == Path.AltDirectorySeparatorChar)
-                    chars[i] = PathSeparator;
+                char* pResult = pChar; // fixed pointers are readonly
+                
+                // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
+                // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
+                // explicitly trying to standardize to '/'
+                for (int i = 0; i < result.Length; i++)
+                {
+                    if (pResult[i] == Path.DirectorySeparatorChar || pResult[i] == Path.AltDirectorySeparatorChar)
+                        pResult[i] = PathSeparator;
+                }
             }
-
-            return new string(chars);
+            
+            return result;
```

Comment:
As @Maxwe11 stated, we don't want to be doing String mutation like this, and @justinvp already changed this implementation in https://github.com/dotnet/corefx/pull/2675.  Please revert the changes to this file.